{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies ‘countrycode’, ‘corrplot’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "\n",
    "#BiocManager::install(\"treeio\")\n",
    "#BiocManager::install(\"ggtree\")\n",
    "#BiocManager::install(\"Biostrings\")\n",
    "#BiocManager::install(\"XVector\")\n",
    "\n",
    "#install.packages(\"tidyverse\")\n",
    "#install.packages(\"ggplot2\")\n",
    "#install.packages(\"ggstance\")\n",
    "#install.packages(\"ggmsa\")\n",
    "#install.packages(\"seqmagick\")\n",
    "#install.packages(\"shiny\")\n",
    "#install.packages(\"DT\")\n",
    "#install.packages(\"cowplot\")\n",
    "#install.packages(\"ggplotify\")\n",
    "#install.packages(\"shinydashboard\")\n",
    "#install.packages(\"shinydashboardPlus\")\n",
    "#install.packages(\"fontawesome\")\n",
    "#install.packages(\"UpSetR\")\n",
    "#install.packages(\"ggvenn\")\n",
    "###install.packages(\"plotly\")\n",
    "#install.packages(\"shinyjs\")\n",
    "install.packages(\"echarts4r\")\n",
    "###install.packages(\"rbokeh\")\n",
    "###install.packages(\"ggiraph\")\n",
    "###install.packages(\"VennDiagram\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library(treeio)\n",
    "library(ggtree)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(ggstance)\n",
    "library(Biostrings)\n",
    "library(ggmsa)\n",
    "library(seqmagick)\n",
    "library(XVector)\n",
    "library(gtable)\n",
    "library(grid)\n",
    "library(shiny)\n",
    "library(DT)\n",
    "library(cowplot)\n",
    "library(ggplotify)\n",
    "library(shinydashboard)\n",
    "library(shinydashboardPlus)\n",
    "library(fontawesome)\n",
    "library(UpSetR)\n",
    "library(ggvenn)\n",
    "###library(plotly)\n",
    "library(shinyjs)\n",
    "library(echarts4r)\n",
    "###library(ggiraph)\n",
    "###library(rbokeh)\n",
    "###library(VennDiagram)\n",
    "###"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ok so let's build a dashboard web app which displays some trees\n",
    "## overview: 4 parts - preprocessing, ui, server, app calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## define my two classes - defining the A2TEA HYPOTHESES object ;D\n",
    "# class for the expanded_OG - containing all different types of data we have on it\n",
    "setClass(\"expanded_OG\", slots=list(genes=\"spec_tbl_df\", \n",
    "                                  fasta_files=\"list\", \n",
    "                                  msa=\"AAMultipleAlignment\", \n",
    "                                  tree=\"phylo\"))\n",
    "\n",
    "# class for the hypotheses\n",
    "setClass(\"hypothesis\", slots=list(description=\"character\", \n",
    "                                  number=\"character\",\n",
    "                                  expanded_in =\"character\", \n",
    "                                  compared_to=\"character\", \n",
    "                                  expanded_OGs=\"list\",\n",
    "                                  species_tree=\"phylo\"))\n",
    "\n",
    "\n",
    "#load the A2TEA input\n",
    "#load(\"/home/tyll/Desktop/PhD/A2TEA/shiny/example_trees/A2TEA_finished.RData\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ui"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "header <- shinydashboardPlus::dashboardHeader(\n",
    "  title = \"A2TEA\",\n",
    "  tags$li(\n",
    "      class = \"dropdown\",\n",
    "# the following CSS keeps the header nice looking and on one line\n",
    "# the CSS at the beginning of the sidebar definitons is also required for this\n",
    "# sizes of individual elements and distances are important however the most important paramter for this to work is \"height\"\n",
    "# this is set to 60px in all elements and ensures nice design \n",
    "## I'll probably have to adapt this a bit if/when I have a nice logo for A2TEA\n",
    "## e.g. a hex design - https://github.com/GuangchuangYu/hexSticker\n",
    "            tags$style(\".main-header {max-height: 60px;}\"),\n",
    "            tags$style(\".main-header .logo {height: 60px;font-size:26px;font-weight:bold;padding:10px;}\"),\n",
    "            tags$style(\".sidebar-toggle {height: 60px; padding-top: 12px !important;font-size: 26px;}\"),\n",
    "            tags$style(\".navbar {min-height:60px !important}\"),\n",
    "            tags$style(\".form-control {padding: 16px 10px;}\")),\n",
    "      # add a github button in the header - to the repo or my profile\n",
    "      tags$li(a(onclick = \"onclick = window.open('https://github.com/tgstoecker')\",\n",
    "                                    href = NULL,\n",
    "                                    icon(\"github\"),\n",
    "                                    title = \"GitHub\",\n",
    "                                    style = \"cursor: pointer; font-size: 32px; height: 60px;\"),\n",
    "                                    class = \"dropdown\"),\n",
    "          # add a twitter button in the header - to my account\n",
    "      tags$li(a(onclick = \"onclick = window.open('https://twitter.com/tgstoecker')\",\n",
    "                                    href = NULL,\n",
    "                                    icon(\"twitter\"),\n",
    "                                    title = \"Twitter\",\n",
    "                                    style = \"cursor: pointer; font-size: 32px; height: 60px;\"),\n",
    "                                    class = \"dropdown\")\n",
    "\n",
    ")\n",
    "\n",
    "\n",
    "sidebar <- dashboardSidebar(\n",
    "  # call useShinyjs() in order to use it!\n",
    "  useShinyjs(),\n",
    "#  actionButton(inputId = \"toggle\", \"Hide/show plot\"),\n",
    "  # as explained in the header creation - necessary CSS to guarantee complete vertical alignment\n",
    "  tags$style(\".left-side, .main-sidebar {padding-top: 60px}\"),\n",
    "  # this code hides the upload and select panels when the the sidebar is toggled off\n",
    "  # it keeps the sidebar tabs though ;D\n",
    "  # this is achieved by specifically referencing (and thus hiding) \".form-group\"\n",
    "  # https://www.w3schools.com/css/css_display_visibility.asp\n",
    "  tags$style(\".sidebar-collapse .form-group {display:none;}\"),\n",
    "  tags$style(\".sidebar-collapse .logo {display:none;}\"),\n",
    "  # Add input of A2TE.RData object\n",
    "  fileInput(inputId = \"A2TEA\",\n",
    "                label = \"Upload A2TEA.RData file:\",\n",
    "                multiple = FALSE,\n",
    "                accept = \".RData\"),\n",
    "  # Add selectors for Hypothesis & HOG\n",
    "  uiOutput('select_Hypothesis'),\n",
    "#  uiOutput('select_HOG'),\n",
    "  sidebarMenu(\n",
    "    menuItem(\"General\", tabName = \"general\", icon = icon(\"th\")),\n",
    "    menuItem(\"TEA analysis\", icon = icon(\"th\"), tabName = \"tea\"), #badgeLabel = \"new\", badgeColor = \"green\")\n",
    "    menuItem(\"Differential expression\", icon = icon(\"th\"), tabName = \"deg\"),\n",
    "    menuItem(\"Gene set enrichment\", icon = icon(\"th\"), tabName = \"gsea\")\n",
    "  )\n",
    "\n",
    ")\n",
    "\n",
    "body <- dashboardBody(\n",
    "  # call useShinyjs() in order to use it!\n",
    "  useShinyjs(),\n",
    "    \n",
    "  tabItems(\n",
    "    # first tab: general information; species plot, upsetR/venn diagramme\n",
    "    tabItem(tabName = \"general\",\n",
    "            h2(\"General information on the experiment\"),\n",
    "            # row1\n",
    "            fluidRow(\n",
    "              # Column 1\n",
    "              column(width = 7, \n",
    "                     infoBox(\n",
    "                       width = NULL,\n",
    "                       title = \"Inferred general stats\",\n",
    "                       subtitle = \"Lorem ipsum blablabla\",\n",
    "                       fill = TRUE,\n",
    "                       color = \"yellow\"\n",
    "                     )\n",
    "                  ),\n",
    "              # Column 2\n",
    "              column(width = 5,\n",
    "                # in order to use shinyjs on box elements - wrap them in div() expressions\n",
    "                # give the divs their own ids! in order to use shinyjs on the box\n",
    "                shinyjs::hidden(\n",
    "                  div(id = \"id_speciesPlot_outer\",\n",
    "                  box(\n",
    "                    id = \"id_speciesPlot\",\n",
    "                    background =\"yellow\",\n",
    "                    gradient = TRUE,\n",
    "                    width = NULL,\n",
    "                    title = \"Phylogenetic Tree - Orthofinder/STAG/STRIDE\",\n",
    "                    plotOutput('speciesPlot')\n",
    "                  )\n",
    "                  )\n",
    "                )\n",
    "              )\n",
    "            ),\n",
    "            \n",
    "            # row 2\n",
    "            fluidRow(\n",
    "              # Column 1\n",
    "              column(width = 3,\n",
    "                     shinyjs::hidden(\n",
    "                       div(id = \"id_vennPlot_outer\",\n",
    "                         box(\n",
    "                           background =\"yellow\",\n",
    "                           gradient = TRUE,\n",
    "                           width = NULL,\n",
    "                           title = \"VennDiagram - Orthologous Groups\",\n",
    "                           plotOutput('vennPlot')\n",
    "                         )\n",
    "                       )\n",
    "                     )  \n",
    "                    ),\n",
    "              # Column 2\n",
    "              column(width = 4,\n",
    "                shinyjs::hidden(\n",
    "                  div(id = \"id_upsetRPlot_outer\",\n",
    "                    box(\n",
    "                      background =\"yellow\",\n",
    "                      gradient = TRUE,\n",
    "                      width = NULL,\n",
    "                      title = \"UpsetR Plot - Orthologous Groups\",\n",
    "                      plotOutput('upsetRPlot')\n",
    "                    )\n",
    "                  )\n",
    "                )\n",
    "              ),\n",
    "              # Column 3\n",
    "              column(width = 5)\n",
    "            )\n",
    "           ),\n",
    "      \n",
    "    # second tab: Trait-specific evolutionary adaptation analysis\n",
    "    tabItem(tabName = \"tea\",\n",
    "            h2(\"Trait-specific evolutionary adaption analysis\"),\n",
    "            uiOutput('select_HOG'),\n",
    "            # select tree layout\n",
    "            selectInput(\n",
    "              inputId = \"tree_layout\",\n",
    "              label = \"Tree layout\",\n",
    "              choices = c('rectangular', 'dendrogram', 'slanted', 'ellipse', 'roundrect', 'fan', 'circular', 'inward_circular', 'radial', 'equal_angle', 'daylight', 'ape')\n",
    "            ),\n",
    "            # turn off/on branch_length option\n",
    "            selectInput(\n",
    "              inputId = \"branch_length\",\n",
    "              label = \"Branch length\",\n",
    "              choices = c('branch.length', 'none')\n",
    "            ),\n",
    "            # turn off/on branch_length option\n",
    "            sliderInput(min = 0.1, max = 2.5, value = 1.0,\n",
    "              inputId = \"tree_size\",\n",
    "              label = \"Tree Size\"\n",
    "            ),\n",
    "            numericInput(inputId = \"tree_height\", \"height\", value = 1200),\n",
    "#            sliderInput(inputId = \"tree_width\", \"width\", min = 100, max = 500, value = 250),\n",
    "            selectInput(inputId = \"tree_plot_legend_pos\", \"Legend position\", choices = c(\"left\", \"top\", \"right\", \"bottom\", \"none\")),\n",
    "            # first row\n",
    "            fluidRow(\n",
    "              # row 1 - column 1\n",
    "              column(width = 4, \n",
    "              ),\n",
    "              # row 1 - column 2\n",
    "              column(width = 8, \n",
    "                shinyjs::hidden(\n",
    "                  div(id = \"id_treePlot_outer\",\n",
    "                  # https://stackoverflow.com/questions/49129194/r-shiny-dynamic-box-height\n",
    "                  # height: XX, where XX can be any CSS unit (I personally prefer viewpoint ratios, e.g. height: 33vh will make a box as high as one third of a screen, regardless of screen resolution), sets the height of your box;\n",
    "                  # overflow-y: auto adds a vertical scrollbar if required.\n",
    "                  # With this approach, when a user resizes the screen all boxes maintain their equal heights.\n",
    "                    box(style = \"height: 1000px; overflow-y: auto; overflow-x: auto\",\n",
    "                      background =\"yellow\",\n",
    "                      gradient = TRUE,\n",
    "                      width = NULL,\n",
    "                      title = \"Tree Plot - Orthologous Group\",\n",
    "                      plotOutput('treePlot')\n",
    "                      )\n",
    "                    )                  \n",
    "                )\n",
    "              )\n",
    "            ),\n",
    "#            plotOutput('treePlot')\n",
    "           ),\n",
    "    tabItem(tabName = \"deg\",\n",
    "            h2(\"Differential expression analysis\"),\n",
    "           ),\n",
    "    tabItem(tabName = \"gsea\",\n",
    "            h2(\"Gene-set enrichment analysis\")\n",
    "           )\n",
    "  )\n",
    ")\n",
    "\n",
    "ui <- shinydashboardPlus::dashboardPage(\n",
    "        skin = \"yellow\",\n",
    "        header = header,\n",
    "        # sidebar should close completely\n",
    "        sidebar = sidebar,\n",
    "        body = body\n",
    "      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [],
   "source": [
    "?ggtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Listening on http://127.0.0.1:6268\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "server <- function(input, output) {\n",
    "  \n",
    "  # change maximum file upload size = default 100Mb\n",
    "  options(shiny.maxRequestSize=100*1024^2)\n",
    "\n",
    "  ## event reactive expression for loading the data once uploaded by the user\n",
    "  # hypotheses.tsv (simple df with)\n",
    "  hypotheses_tsv <- eventReactive(input$A2TEA, {\n",
    "    t = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t)\n",
    "    get(\"hypotheses\", envir=t)\n",
    "  })\n",
    "  # core A2TEA object\n",
    "  HYPOTHESES.a2tea <- eventReactive(input$A2TEA, {\n",
    "    t = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t)\n",
    "    get(\"HYPOTHESES.a2tea\", envir=t)\n",
    "  })\n",
    "  # Gene level diff. exp. table\n",
    "  HOG_DE.a2tea <- eventReactive(input$A2TEA, {\n",
    "    t = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t)\n",
    "    get(\"HOG_DE.a2tea\", envir=t)\n",
    "  })\n",
    "  # List of HOGs\n",
    "  HOG_level_list <- eventReactive(input$A2TEA, {\n",
    "    t = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t)\n",
    "    get(\"HOG_level_list\", envir=t)\n",
    "  })\n",
    "    \n",
    "\n",
    "  ## observe upload of data and only then show boxes in which the plots are displayed\n",
    "  observeEvent(input$A2TEA, {\n",
    "    shinyjs::showElement(\"id_speciesPlot_outer\", anim = TRUE, animType = \"slide\", time = 1)\n",
    "    shinyjs::showElement(\"id_vennPlot_outer\", anim = TRUE, animType = \"slide\", time = 1)\n",
    "    shinyjs::showElement(\"id_upsetRPlot_outer\", anim = TRUE, animType = \"slide\", time = 1)\n",
    "    # TEA analysis tab\n",
    "    shinyjs::showElement(\"id_treePlot_outer\", anim = TRUE, animType = \"slide\", time = 1)\n",
    "  })\n",
    "    \n",
    "  ## rendering of UI elements for the sidebar\n",
    "  # choice of hypothesis to be displayed\n",
    "    output$select_Hypothesis = renderUI({\n",
    "      selectInput(inputId = 'select_Hypothesis_server', \n",
    "                  label = 'Select Hypothesis:', \n",
    "                  choices = hypotheses_tsv()$name\n",
    "                 )\n",
    "    })\n",
    "  \n",
    "  # a dropdown list of the expanded HOGs of the current hypothesis\n",
    "  # -> observe choice of HOG; e.g. for building trees\n",
    "  # first create reactive list of all exp. HOGs \n",
    "  hypothesisExpOGs <- reactive({\n",
    "    # require the input$select_Hypothesis_server to be present, since otherwise for a brief second an error is thrown\n",
    "    req(input$select_Hypothesis_server)\n",
    "    vars <- all.vars(\n",
    "      parse(\n",
    "        text = names(HYPOTHESES.a2tea()[[hypotheses_tsv()$hypothesis[hypotheses_tsv()$name==input$select_Hypothesis_server]]]@expanded_OGs)\n",
    "      )\n",
    "    )    \n",
    "    vars <- as.list(vars)\n",
    "    return(vars)    \n",
    "  })\n",
    "    \n",
    "  output$select_HOG = renderUI({     \n",
    "    selectInput('select_HOG_server', 'Select HOG', hypothesisExpOGs())\n",
    "  })\n",
    "\n",
    "\n",
    "########################\n",
    "# General (start) page\n",
    "########################\n",
    "\n",
    "  # creating the species tree plot\n",
    "  # make renderUI element reactive\n",
    "  # access to element based on name in first element selectInput('')\n",
    "  output$speciesPlot <- renderPlot({\n",
    "    req(input$select_Hypothesis_server)\n",
    "    ggtree(HYPOTHESES.a2tea()[[hypotheses_tsv()$hypothesis[hypotheses_tsv()$name==input$select_Hypothesis_server]]]@species_tree)\n",
    "  })\n",
    "\n",
    "  ## Creating upsetR and venn plots\n",
    "  # based on hypothesis/ when it is changed, create specific list with incl. genotypes/species\n",
    "  genotypeList <- eventReactive(input$select_Hypothesis_server, {\n",
    "      \n",
    "    workset <- HOG_level_list()[[hypotheses_tsv()$hypothesis[hypotheses_tsv()$name==input$select_Hypothesis_server]]] %>% \n",
    "                 select(HOG, ends_with(\"_total\")) \n",
    "    # add this point we can easily compute the number of species/ecotypes etc. \n",
    "    # #species = (columns of current workset - 1) \n",
    "    species_num <- ncol(workset) - 1\n",
    "    # change 0 to NAs since we can more easily filter on that\n",
    "    is.na(workset) <- workset==0\n",
    "    # remove '_total' from column names\n",
    "    #str_remove(colnames(workset), \"_total\")\n",
    "    workset <- workset %>% setNames(names(workset) %>% stringr::str_replace(\"_total\",\"\"))\n",
    "    \n",
    "    genotypeList <- list()\n",
    "    for (j in 1:species_num) {\n",
    "      assign(paste0(\"Genotype\", j), as.vector(simplify_all(workset[!is.na(workset[j+1]),\"HOG\"])[[1]]))\n",
    "      subList <- list(get(paste0(\"Genotype\", j)))\n",
    "      names(subList) <- colnames(workset)[j+1]\n",
    "      genotypeList <- c(genotypeList, subList)\n",
    "    }\n",
    "    # return genotypeList\n",
    "    genotypeList\n",
    "  })\n",
    "\n",
    "  # create Venn diagram\n",
    "  output$vennPlot <- renderPlot({\n",
    "    ggvenn(genotypeList(), \n",
    "           stroke_linetype = 2, \n",
    "           text_size = 5, \n",
    "           set_name_color = \"navy\",\n",
    "           stroke_color = \"yellow3\",\n",
    "           stroke_size = 0.8, \n",
    "           set_name_size = 6, \n",
    "           show_percentage = TRUE, \n",
    "           fill_color = c(\"yellow\", \"purple\", \"papayawhip\", \"red4\")\n",
    "          )\n",
    "  })\n",
    "    \n",
    "  # create upsetR plot \n",
    "  output$upsetRPlot <- renderPlot({\n",
    "    req(input$select_Hypothesis_server)\n",
    "    upset(fromList(genotypeList()), \n",
    "          order.by = \"freq\",\n",
    "          empty.intersections = \"on\",\n",
    "          point.size = 3.5, \n",
    "          line.size = 2, \n",
    "          mainbar.y.label = \"HOG Intersections\", \n",
    "          text.scale = c(2, 1.3, 1, 1, 2, 2),\n",
    "          main.bar.color = \"navy\",\n",
    "          matrix.color = \"navy\",\n",
    "          sets.bar.color = \"navy\"\n",
    "         )\n",
    "    })\n",
    "        \n",
    "    \n",
    "      \n",
    "########################\n",
    "# Tea analysis \n",
    "########################\n",
    "  # read in tree choice reactively\n",
    "  # using eventReactive() and requiring the input$select_HOG_server to change leads to this specifically being the trigger\n",
    "  # with only the basic reactive() all upstream events influence this and this leads to warning messages;\n",
    "  # specifically the class of the tree object isn't found immediately\n",
    "  # since select_HOG_server is dependent on the reactive select_Hypothesis_server we circumvent this here\n",
    "  # changing HOG no problem, changing Hypothesis sets HOG to index 1 and is then evaluated -> no warnings and a clean App\n",
    "  tree_choice <- eventReactive(input$select_HOG_server, {\n",
    "    HYPOTHESES.a2tea()[[hypotheses_tsv()$hypothesis[hypotheses_tsv()$name==input$select_Hypothesis_server]]]@expanded_OGs[[input$select_HOG_server]]@tree\n",
    "  })\n",
    "    \n",
    "    \n",
    "\n",
    "  # creating the tree expansion plot\n",
    "  # make renderUI element reactive\n",
    "  # access to element based on name in first element selectInput('')\n",
    "  #output$\n",
    "    output$treePlot <- renderPlot(\n",
    "#    width = function() input$tree_width,\n",
    "    height = function() input$tree_height,\n",
    "                                {\n",
    "    #    width = \"auto\",\n",
    "    req(input$select_Hypothesis_server)\n",
    "    req(input$select_HOG_server)\n",
    "    \n",
    "    tree <- tree_choice() \n",
    "    info <- HOG_DE.a2tea()\n",
    "      #ggtree(tree)\n",
    "    \n",
    "    # rlang package has the function is_empty()\n",
    "    # with it we can test for \"character(0)\"\n",
    "    el <- list()\n",
    "\n",
    "    for (i in 1:length(tree$tip.label)) {\n",
    "        if (is_empty(info$HOG[info$gene == tree$tip.label[i]]) == FALSE) {\n",
    "           first <- info$HOG[info$gene == tree$tip.label[i]]\n",
    "           el <- c(el, list(filler = tree$tip.label[i]))\n",
    "           names(el)[i] <- first\n",
    "        }\n",
    "        else {\n",
    "            el <- c(el, list(singleton = tree$tip.label[i]))      \n",
    "        }\n",
    "    }\n",
    "\n",
    "    # reduce the list to unique tags\n",
    "    el_split <- sapply(unique(names(el)), function(x) unname(unlist(el[names(el)==x])), simplify=FALSE)\n",
    "\n",
    "    p_OTU <- ggtree(tree, \n",
    "                    layout = input$tree_layout,\n",
    "                    branch.length = input$branch_length,\n",
    "                    size = input$tree_size)\n",
    "  \n",
    "    p <- groupOTU(p_OTU, el_split, 'HOG') + \n",
    "                       aes(color=HOG) +\n",
    "                       geom_tiplab(aes(color=HOG)) +\n",
    "#                       theme(legend.position=\"right\") +\n",
    "                       coord_cartesian(clip = 'off') + \n",
    "                       theme_tree2(plot.margin=margin(6, 120, 6, 6), legend.position= input$tree_plot_legend_pos)\n",
    "                       #scale_color_manual(values=cols,\n",
    "                       #                   na.translate=TRUE,\n",
    "                       #                   na.value = \"black\") +\n",
    "                       #geom_tippoint(aes(color=HOG), align=F, offset = 0.3, alpha=1)\n",
    "                       #geom_tiplab(\"ddkdkd\")\n",
    "                       \n",
    "    return(p)\n",
    "    #height = exprToFunction(as.numeric(input$dimension[2]) *  1/ 2  * 1 / 3)\n",
    "    #p + lims(x = c(0, max(p$data$x) * input$subtree_width_multiply))\n",
    "    #p\n",
    "                       \n",
    "  })\n",
    "    \n",
    "                                    \n",
    "                                    \n",
    "                                    \n",
    "####################\n",
    "# server side row rendering test\n",
    "###############################\n",
    "#  output$row_render_test <- renderUI({\n",
    "#    req(input$select_Hypothesis_server)\n",
    "#    req(input$select_HOG_server)\n",
    "   # output <-  \n",
    "      #fluidRow(\n",
    "   #               column(width = 4,\n",
    "   #                 numericInput(inputId = \"tree_heightNull\", \"tree_heightNull\", value = 3*300),\n",
    "   #                     ),\n",
    "   #               column(width = 8,\n",
    "   #                 box(title = \"dkdkd\", \n",
    "   #                     gradient = TRUE, \n",
    "   #                     background =\"yellow\", \n",
    "   #                     width = NULL,\n",
    "#                        plotOutput(\"treePlot\", height = 3*300)\n",
    "   #                      )\n",
    "   #                     \n",
    "#  })\n",
    "                                    \n",
    "}\n",
    "\n",
    "                              \n",
    "shinyApp(ui, server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [],
   "source": [
    "?ggplot2::theme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load(file = \"A2TEA_finished.RData\")\n",
    "#validObject(HYPOTHESES.a2tea$hypothesis_1@expanded_OGs$N0.HOG0000207@tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating upsetR Plot\n",
    "head(HOG_level_list[[1]])\n",
    "\n",
    "workset <-  HOG_level_list[[1]]\n",
    "# drop columns that are unnecessary\n",
    "workset <- workset %>% select(HOG, ends_with(\"_total\")) \n",
    "# add this point we can easily compute the number of species/ecotypes etc. \n",
    "# #species = (columns of current workset - 1) \n",
    "species_num <- ncol(workset) - 1\n",
    "# change 0 to NAs since we can more easily filter on that\n",
    "is.na(workset) <- workset==0\n",
    "# remove '_total' from column names\n",
    "#str_remove(colnames(workset), \"_total\")\n",
    "workset <- workset %>% setNames(names(workset) %>% stringr::str_replace(\"_total\",\"\"))\n",
    "\n",
    "head(workset)\n",
    "\n",
    "genotypeList <- list()\n",
    "for (j in 1:species_num) {\n",
    "    assign(paste0(\"Genotype\", j), as.vector(simplify_all(workset[!is.na(workset[j+1]),\"HOG\"])[[1]]))\n",
    "    subList <- list(get(paste0(\"Genotype\", j)))\n",
    "    names(subList) <- colnames(workset)[j+1]\n",
    "    genotypeList <- c(genotypeList, subList)\n",
    "}\n",
    "\n",
    "#head(genotypeList)\n",
    "\n",
    "#using UpSetR\n",
    "upsetOrtho <- upset(fromList(genotypeList), \n",
    "                    order.by = \"freq\",\n",
    "                    empty.intersections = \"on\",\n",
    "                    point.size = 3.5, \n",
    "                    line.size = 2, \n",
    "                    mainbar.y.label = \"HOG Intersections\", \n",
    "                    text.scale = c(2, 1.3, 1, 1, 2, 2),\n",
    "                    main.bar.color = \"navy\",\n",
    "                    matrix.color = \"navy\",\n",
    "                    sets.bar.color = \"navy\"\n",
    "                   )\n",
    "\n",
    "upsetOrtho\n",
    "\n",
    "ggvenn(genotypeList, stroke_linetype = 2, text_size = 5, set_name_color = \"navy\",\n",
    "       stroke_color = \"yellow3\",\n",
    "       stroke_size = 0.8, \n",
    "       set_name_size = 6, \n",
    "       show_percentage = TRUE, \n",
    "       fill_color = c(\"yellow\", \"purple\", \"papayawhip\", \"red4\")\n",
    "      )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## UpSetR playaround"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# drop columns that are unnecessary and get names of all species columns with total gene count per HOG\n",
    "workset <- HOG_level_list[[1]] %>% select(ends_with(\"_total\")) %>% names() \n",
    "# create copy of HOG_level_list[[1]] to be worked on\n",
    "work_HOG_level_list <- HOG_level_list[[1]]\n",
    "\n",
    "for (i in workset) {\n",
    "    #print(i)\n",
    "    species_name <- i %>% stringr::str_replace(\"_total\",\"\")\n",
    "    #print(species_name)\n",
    "    # create new columns for each species - 0,1 whether they have a gene in each HOG - this is needed or UpSetR\n",
    "    work_HOG_level_list <- mutate(work_HOG_level_list, !!species_name := ifelse(\n",
    "                                                      get(i) > 0, as.integer(1), as.integer(0)\n",
    "                                                         )\n",
    "         )\n",
    "}\n",
    "str(work_HOG_level_list)\n",
    "\n",
    "\n",
    "#subset_work_HOG_level_list <- work_HOG_level_list %>% select(HOG, Arabidopsis_thaliana, Zea_mays, Hordeum_vulgare)\n",
    "#str(subset_work_HOG_level_list)\n",
    "\n",
    "# can't use tibble apparently...\n",
    "# convert to dataframe first\n",
    "df_work_HOG_level_list <- as.data.frame(work_HOG_level_list)\n",
    "str(df_work_HOG_level_list)\n",
    "\n",
    "df_work_HOG_level_list$Zea_mays_sigDE <- as.numeric(df_work_HOG_level_list$Zea_mays_sigDE)\n",
    "df_work_HOG_level_list$Zea_mays_sigDE[is.na(df_work_HOG_level_list$Zea_mays_sigDE)] <- 0\n",
    "#str(df_work_HOG_level_list)\n",
    "\n",
    "\n",
    "#custom func to colour if sig DE genes from species A. thaliana\n",
    "Myfunc <- function(row, num) {\n",
    "    data <- (row[\"Hordeum_vulgare_sigDE\"] > num)\n",
    "}\n",
    "\n",
    "#df_work_HOG_level_list %>% filter(!is.na(Zea_mays_sigDE)) %>% head()\n",
    "\n",
    "upset(df_work_HOG_level_list,\n",
    "      sets = c(\"Arabidopsis_thaliana\", \"Zea_mays\", \"Hordeum_vulgare\"),\n",
    "      order.by = \"freq\",\n",
    "      empty.intersections = \"on\",\n",
    "      point.size = 3.5, \n",
    "      line.size = 2, \n",
    "      mainbar.y.label = \"HOG Intersections\", \n",
    "      text.scale = c(2, 1.3, 1, 1, 2, 2),\n",
    "      sets.bar.color = \"navy\",\n",
    "      queries = list(list(query = Myfunc, params=list(3), color = \"green\", active = T))\n",
    "     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head(df_work_HOG_level_list)\n",
    "#names(df_work_HOG_level_list)\n",
    "#hist(df_work_HOG_level_list$Zea_mays_sigDE)\n",
    "#df_work_HOG_level_list[!is.na(df_work_HOG_level_list$Zea_mays_sigDE),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#movies <- read.csv(system.file(\"extdata\", \"movies.csv\", package = \"UpSetR\"), \n",
    "#    header = T, sep = \";\")\n",
    "str(movies)\n",
    "\n",
    "Myfunc <- function(row, rating) {\n",
    "    data <- (row[\"AvgRating\"] > rating)\n",
    "}\n",
    "\n",
    "Myfunc2 <- function(row, rating) {\n",
    "    data <- (row[\"AvgRating\"] < rating)\n",
    "}\n",
    "\n",
    "upset(movies, sets = c(\"Adventure\", \"Action\", \"SciFi\"),\n",
    "     queries = list(list(query = Myfunc, params = 2.5, color = \"blue\", active = T))#,\n",
    "#                   list(query = Myfunc2, params = 2.5, color = \"green\", active = T))\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ui"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define UI for tree app ----\n",
    "\n",
    "#hypothesesList <- as.list(hypotheses$hypothesis)\n",
    "#names(hypothesesList) <- hypotheses$name\n",
    "#hypothesesList\n",
    "\n",
    "\n",
    "# Allow specific errors to be displayed on screen, instead of displaying a generic error\n",
    "options(shiny.sanitize.errors = FALSE)\n",
    "\n",
    "ui <- pageWithSidebar(\n",
    "  \n",
    "  # App title ----\n",
    "  headerPanel(\"Tree building\"),\n",
    "  \n",
    "  # Sidebar panel for inputs ----\n",
    "  sidebarPanel(\n",
    "  \n",
    "      fileInput(inputId = \"A2TEA\",\n",
    "                label = \"Upload A2TEA.RData file:\",\n",
    "                multiple = FALSE,\n",
    "                accept = \".RData\"),\n",
    "      \n",
    "        # Input: alternative way which leads to problems...\n",
    "#      selectInput(\"select\",\n",
    "#                    label = h3(\"Select hypothesis:\"),\n",
    "#                    \"\"), \n",
    "\n",
    "      uiOutput('select_Hypothesis'),\n",
    "      uiOutput('select_HOG')\n",
    "      \n",
    "  ),\n",
    "  \n",
    "  # Main panel for displaying outputs ----\n",
    "  mainPanel(\n",
    "      h3(textOutput(\"text\")),\n",
    "      plotOutput(\"speciesTree\"),\n",
    "      plotOutput(\"VennUpSet_plot\"),\n",
    "      DT::dataTableOutput(\"HOG_DE_DT\"),\n",
    "      DT::dataTableOutput(\"HOG_level\"),\n",
    "      plotOutput(\"recFacetPlot\"),\n",
    "  )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#increasing maximum file size for upload!\n",
    "#https://stackoverflow.com/questions/18037737/how-to-change-maximum-upload-size-exceeded-restriction-in-shiny-and-save-user\n",
    "options(shiny.maxRequestSize=50*1024^2)\n",
    "\n",
    "myDataFrame <- data.frame(names=c(\"\"))\n",
    "#rm(myDataFrame)\n",
    "\n",
    "server <- function(input, output, session) {\n",
    "    \n",
    "\n",
    "#read-in hypotheses.tsv reactively\n",
    "    tsv <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t)\n",
    "    d <- get(\"hypotheses\", envir=t)\n",
    "     }\n",
    "     d\n",
    "   })  \n",
    "    \n",
    "#    observe({\n",
    "#       updateSelectInput(session, \"select\",\n",
    "#                         label = \"Select Hypothesis:\",\n",
    "#                         choices = tsv()$name,\n",
    "#                         selected = tsv()$name[1])\n",
    "# })\n",
    "    \n",
    "    \n",
    "#    output$hypothesesList <- renderText({\n",
    "#        tsv()$name\n",
    "#    })\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "#    tsv2 <- reactive({\n",
    "#       vars <- all.vars(parse(text = names(HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs)))\n",
    "#       vars <- as.list(vars)\n",
    "#       return(vars)\n",
    "    \n",
    "    \n",
    "    output$select_Hypothesis = renderUI({\n",
    "      selectInput('select_Hypothesis2', 'Select Hypothesis:', tsv()$name)\n",
    "    })\n",
    "\n",
    "# also read-in the Ortho Venn/UpSet plots \n",
    "    Ortho_intersect_plots <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t5 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t5)\n",
    "    d <- get(\"Ortho_intersect_plots\", envir = t5)\n",
    "     }\n",
    "     d\n",
    "   }) \n",
    "\n",
    "    #plot the Ortho Venn/UpSet plots\n",
    "    output$VennUpSet_plot <- renderPlot({\n",
    "        Ortho_intersect_plots()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]\n",
    "    })\n",
    "\n",
    "    \n",
    "# also read-in the DE genes table \n",
    "    HOG_DE <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t2 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t2)\n",
    "    d <- get(\"HOG_DE.a2tea\", envir = t2)\n",
    "     }\n",
    "     d\n",
    "   }) \n",
    "\n",
    "\n",
    "# also read-in custom A2TEA object\n",
    "    HYPOTHESES <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t3 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t3)\n",
    "    d <- get(\"HYPOTHESES.a2tea\", envir = t3)\n",
    "     }\n",
    "     d\n",
    "   })  \n",
    "\n",
    "# buld species tree\n",
    "    #plot the Ortho Venn/UpSet plots\n",
    "    output$speciesTree <- renderPlot({\n",
    "       st <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@species_tree\n",
    "       ggtree(st, layout = 'rectangular', branch.length='yes', size=1.5) +\n",
    "           geom_tiplab(align=F, offset = 0.1, alpha=1)\n",
    "            \n",
    "    })\n",
    "    \n",
    "    \n",
    "# also read-in hypothesis specific HOG level table - HOG_level_list\n",
    "    HOG_level <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t4 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t4)\n",
    "    d <- get(\"HOG_level_list\", envir = t4)\n",
    "     }\n",
    "     d\n",
    "   })  \n",
    "    \n",
    "    \n",
    "# normal table chokes on large size of df\n",
    "# therefore usage of DT::renderDataTable\n",
    "    output$HOG_DE_DT <- DT::renderDataTable({\n",
    "        HOG_DE()\n",
    "  })\n",
    "    \n",
    "    \n",
    "#example access to hypothesis number - useful for access \n",
    "    hypothesisNUM <- reactive({\n",
    "#       HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@number\n",
    "        HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@number\n",
    "    })\n",
    "   \n",
    "     output$text <- renderText({\n",
    "        hypothesisNUM()\n",
    "    })    \n",
    "        \n",
    "        \n",
    "\n",
    "        \n",
    "# I can make use of the hypothesis number previously created\n",
    "    output$HOG_level <- DT::renderDataTable({\n",
    "#        HOG_level()[[tsv()$hypothesis[tsv()$name==input$select]]]\n",
    "                HOG_level()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]\n",
    "  })\n",
    "             \n",
    "     \n",
    "        \n",
    "\n",
    "\n",
    "# observe choice of HOG; e.g. for building trees\n",
    "# first create reactive list of all exp. HOGs \n",
    "# due to my custom object this needs the following trickery: \n",
    "# all.vars(parse(text = names(....\n",
    "hypothesisExpOGs <- reactive({\n",
    "#       vars <- all.vars(parse(text = names(HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs)))\n",
    "       vars <- all.vars(parse(text = names(HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@expanded_OGs)))    \n",
    "       vars <- as.list(vars)\n",
    "       return(vars)    \n",
    "    })\n",
    "    \n",
    "# actually render the UI element - a dropdown list of the expanded HOGs of the current hypothesis\n",
    "    output$select_HOG = renderUI({\n",
    "      selectInput('select_HOG2', 'Select HOG', hypothesisExpOGs())\n",
    "    })\n",
    "    \n",
    "# creating the rectangular A2TEA triple facet plot\n",
    "\n",
    "# make renderUI element reactive\n",
    "# access to element based on name in first element selectInput('')\n",
    "    choice <- reactive({input$select_HOG2})\n",
    "    \n",
    "output$recFacetPlot <- renderPlot({\n",
    "\n",
    "    #### building the rectangular facet plot - HOG tree, logFC & MSA\n",
    "                     \n",
    "#    tree <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs[[choice()]]@tree\n",
    "    tree <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@expanded_OGs[[choice()]]@tree\n",
    "    info <- HOG_DE()\n",
    "\n",
    "\n",
    "# extracted from colourblind - removed black... how to deal with NAs..\n",
    "# also a problem if more than given colours are supposed to be displayed error....\n",
    "# could perform check of number of HOGs and switch between an optimized colour scale with predefined colours\n",
    "# to an open-ended/automatic colour scheme if set number is exceeded by choice of HOG\n",
    "    cols <- c(\"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\", \"#D55E00\", \"#CC79A7\", \"lightblue\", \"darkred\",\n",
    "              \"lightblue\", \"darkred\", \"pink\", \"cyan\", \"grey\", \"violet\", \"magenta\", \"gold\", \n",
    "              \"darkgreen\", \"darkblue\")\n",
    "\n",
    "\n",
    "# rlang package has the function is_empty()\n",
    "# with it we can test for \"character(0)\"\n",
    "    el <- list()\n",
    "\n",
    "    for (i in 1:length(tree$tip.label)) {\n",
    "        if (is_empty(info$HOG[info$gene == tree$tip.label[i]]) == FALSE) {\n",
    "           first <- info$HOG[info$gene == tree$tip.label[i]]\n",
    "           el <- c(el, list(filler = tree$tip.label[i]))\n",
    "           names(el)[i] <- first\n",
    "        }\n",
    "        else {\n",
    "            el <- c(el, list(singleton = tree$tip.label[i]))      \n",
    "        }\n",
    "    }\n",
    "\n",
    "# reduce the list to unique tags\n",
    "    el_split <- sapply(unique(names(el)), function(x) unname(unlist(el[names(el)==x])), simplify=FALSE)\n",
    "\n",
    "    p_OTU <- ggtree(tree,\n",
    "                    layout = 'rectangular',\n",
    "                    branch.length='none',\n",
    "                    size=1.5)\n",
    "\n",
    "    p <- groupOTU(p_OTU, el_split, 'HOG') + \n",
    "                       aes(color=HOG) +\n",
    "                       geom_tippoint(aes(color=HOG)) +\n",
    "                       scale_color_manual(values=cols,\n",
    "                                          na.translate=TRUE,\n",
    "                                          na.value = \"black\") +\n",
    "                       geom_tiplab(aes(color=HOG), \n",
    "                                   align=F, \n",
    "                                   offset = 0.3,\n",
    "                                   alpha=1)\n",
    "                                   \n",
    "\n",
    "# this is such a convenient way to do this ;D + we also drop the columns we don't need\n",
    "    d <- filter(info, gene %in% tree$tip.label) %>% \n",
    "             select(-c(species, baseMean, HOG, lfcSE, stat, pvalue, padj))\n",
    "\n",
    "    lines=data.frame(y = c(-2,-1,1,2), .panel='log2FC')\n",
    "\n",
    "                   \n",
    "# xlim set x axis limits for only Tree panel\n",
    "# xlim tree calculated as number of tips tips?\n",
    "    p2 <- facet_plot(p + xlim_tree(16), \n",
    "                     panel = 'log2FC', \n",
    "                     data = d, \n",
    "                     geom = geom_barh, \n",
    "                     mapping = aes(x = log2FoldChange),\n",
    "                     stat='identity',\n",
    "                 #inherit.aes = TRUE,\n",
    "                     color='firebrick') + \n",
    "                       theme_tree2(plot.margin=margin(0, 0, 0, 0)) + \n",
    "                       geom_vline(data=lines, \n",
    "                                  aes(xintercept=y), \n",
    "                                  linetype = \"dashed\")\n",
    "                   \n",
    "                   \n",
    "# get the msa and assign it to x (for now...)\n",
    "# set msa range of amino acids to be displayed - make this dynamcic!\n",
    "#    x <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs[[choice()]]@msa\n",
    "    x <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@expanded_OGs[[choice()]]@msa\n",
    "    data <- tidy_msa(x, start = 1, end = 598)\n",
    "\n",
    "\n",
    "# use font to display amino acid code, eg \"helvetical\"\n",
    "p3 <- p2 + geom_facet(geom = geom_msa, \n",
    "               data = data,  \n",
    "               panel = 'msa',\n",
    "               font = NULL, \n",
    "               color = \"Clustal\")\n",
    "                   \n",
    "\n",
    "    gt = ggplot_gtable(ggplot_build(p3))\n",
    "#gt$layout$l[grep('background', gt$layout$name)]\n",
    "#gt$widths[1] = 20*gt$widths[1] # in this case it was colmun 5 - double the width\n",
    "    gt$widths[5] = 2*gt$widths[5] # in this case it was colmun 5 - double the width\n",
    "    gt$widths[7] = 2*gt$widths[7] # in this case it was colmun 7 - double the width\n",
    "    gt$widths[9] = 5*gt$widths[9] # in this case it was colmun 9 - double the width\n",
    "\n",
    "# this option is nice for jupyter but I have to see what will be possible in shiny                   \n",
    "    options(repr.plot.width=15, repr.plot.height=8)\n",
    "    grid.draw(gt)\n",
    "            \n",
    "            \n",
    "############ ending facet plot code ############\n",
    "            \n",
    "  })\n",
    "        \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### app calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shinyApp(ui, server)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### playground & working examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hypotheses$hypothesis[hypotheses$name==\"Expanded in barley compared to maize\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS( data.frame(names=c(\"Jill\",\"Jane\",\"Megan\")),\"example_trees/myDataFrame.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### completely working mini application showcasing runtime loading of data and reactive dropdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myDataFrame <- data.frame(names=c(\"Tom\",\"Dick\",\"Harry\"))\n",
    "\n",
    "ui <- shinyUI(\n",
    "    fluidPage(\n",
    "        fileInput(\"file1\", \"Choose file to upload\", accept = \".rds\"),\n",
    "        selectInput(\"myNames\",\"Names\", \"\"),\n",
    "        tableOutput(\"contents\")\n",
    "    )\n",
    ")\n",
    "\n",
    "server <- function(input, output, session) {\n",
    "\n",
    "    myData <- reactive({\n",
    "        inFile <- input$file1\n",
    "        if (is.null(inFile)) {\n",
    "            d <- myDataFrame\n",
    "        } else {\n",
    "            d <- readRDS(inFile$datapath)\n",
    "        }\n",
    "        d\n",
    "    })\n",
    "\n",
    "    output$contents <- renderTable({\n",
    "        myData()\n",
    "    })\n",
    "\n",
    "    observe({\n",
    "         updateSelectInput(session, \"myNames\",\n",
    "                           label = \"myNames\",\n",
    "                           choices = myData()$names,\n",
    "                           selected = myData()$names[1])\n",
    "    })\n",
    "\n",
    "}\n",
    "\n",
    "shinyApp(ui, server)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### shiny example of using RData input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The example allows you to upload .RData files.\n",
    "#The approach with load and get allows you to assign the loaded data to a variable name of your choice.\n",
    "#For the matter of the example being \"standalone\" I inserted the top section that stores \n",
    "#two vectors to your disk in order to load and plot them later.\n",
    "\n",
    "\n",
    "# Define two datasets and store them to disk\n",
    "x <- rnorm(100)\n",
    "save(x, file = \"x.RData\")\n",
    "#rm(x)\n",
    "y <- rnorm(100, mean = 2)\n",
    "save(y, file = \"y.RData\")\n",
    "#rm(y)\n",
    "# both\n",
    "save(x, y, file = \"z.RData\")\n",
    "\n",
    "# This function, borrowed from http://www.r-bloggers.com/safe-loading-of-rdata-files/, load the Rdata into a new environment to avoid side effects\n",
    "LoadToEnvironment <- function(RData, env=new.env()) {\n",
    "  load(RData, env)\n",
    "  return(env)\n",
    "}\n",
    "\n",
    "\n",
    "# Define UI\n",
    "ui <- shinyUI(fluidPage(\n",
    "  titlePanel(\".RData File Upload Test\"),\n",
    "  mainPanel(\n",
    "    fileInput(\"file\", label = \"\"),\n",
    "    actionButton(inputId=\"plot\",\"Plot\"),\n",
    "    plotOutput(\"hist\"))\n",
    "  )\n",
    ")\n",
    "\n",
    "# Define server logic\n",
    "server <- shinyServer(function(input, output) {\n",
    "\n",
    "  observeEvent(input$plot,{\n",
    "    if ( is.null(input$file)) return(NULL)\n",
    "    inFile <- input$file\n",
    "    file <- inFile$datapath\n",
    "    # load the file into new environment and get it from there\n",
    "#    name <- load(file, envir = e)\n",
    "#    data <- e[[name]]\n",
    "    # this works but not really what I need\n",
    "#    data <- e[[name[1]]]\n",
    "    t = new.env()\n",
    "    load(file, envir = t)\n",
    "    data <- get(\"y\", envir=t)\n",
    "\n",
    "      \n",
    "      \n",
    "#    observeEvent(input$plot,{\n",
    "#        if ( is.null(input$file)) return(NULL)\n",
    "#        inFile <- input$file\n",
    "#        file <- inFile$datapath\n",
    "#    # Use a reactiveFileReader to read the file on change, and load the content into a new environment\n",
    "#    env <- reactiveFileReader(1000, session, \"z.RData\", LoadToEnvironment)\n",
    "#    # Access the first item in the new environment, assuming that the Rdata contains only 2 items which is a data frame\n",
    "#    env()[[names(env())[2]]]\n",
    "    \n",
    "    \n",
    "    # Plot the data\n",
    "    output$hist <- renderPlot({\n",
    "      hist(data)\n",
    "    })\n",
    "  })\n",
    "})\n",
    "\n",
    "# Run the application \n",
    "shinyApp(ui = ui, server = server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = new.env()\n",
    "name <- load(\"z.RData\", envir = e)\n",
    "#mget(c(\"y\",\"x\"), envir=e)\n",
    "#get returns exactly the object\n",
    "test <- get(\"y\", envir=e)\n",
    "str(test)\n",
    "str(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load(\"example_trees/backup_old_RData//A2TEA_finished.RData\")\n",
    "#str(hypotheses$name)\n",
    "#HOG_DE.a2tea\n",
    "#HYPOTHESES.a2tea\n",
    "#HOG_level_list[[1]][HOG_level_list[[1]]$expansion==\"yes\",]\n",
    "#tt <- as.data.frame(names(HYPOTHESES.a2tea[[1]]@expanded_OGs))\n",
    "#names(tt)\n",
    "#names(tt)[names(tt) == \"names(HYPOTHESES.a2tea[[1]]@expanded_OGs)\"] <- \"HOG\"\n",
    "#names(tt)\n",
    "#tt$HOG\n",
    "#tt[[1]]\n",
    "#?base::as.data.frame\n",
    "\n",
    "\n",
    "load(\"A2TEA_finished.RData\")\n",
    "#str(HOG_level_list)\n",
    "#hypotheses\n",
    "#HYPOTHESES.a2tea$hypothesis_1@description\n",
    "#HYPOTHESES.a2tea()[[hypotheses_tsv()$hypothesis[hypotheses_tsv()$name==input$select_Hypothesis_server]]]@expanded_OGs\n",
    "\n",
    "\n",
    "HYPOTHESES.a2tea[[hypotheses$hypothesis[hypotheses$name==\"Expanded in Arabidopsis compared to Monocots\"]]]@expanded_OGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HYPOTHESES.a2tea[[hypotheses$hypothesis[hypotheses$name==\"Expanded in Arabidopsis compared to Monocots\"]]]@number"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
