{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependency ‘gridGraphics’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "\n",
    "#BiocManager::install(\"treeio\")\n",
    "#BiocManager::install(\"ggtree\")\n",
    "#BiocManager::install(\"Biostrings\")\n",
    "#BiocManager::install(\"XVector\")\n",
    "\n",
    "#install.packages(\"tidyverse\")\n",
    "#install.packages(\"ggplot2\")\n",
    "#install.packages(\"ggstance\")\n",
    "#install.packages(\"ggmsa\")\n",
    "#install.packages(\"seqmagick\")\n",
    "#install.packages(\"shiny\")\n",
    "#install.packages(\"DT\")\n",
    "#install.packages(\"cowplot\")\n",
    "#install.packages(\"ggplotify\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'treeio':\n",
      "  method     from\n",
      "  root.phylo ape \n",
      "\n",
      "treeio v1.13.1  For help: https://yulab-smu.github.io/treedata-book/\n",
      "\n",
      "If you use treeio in published research, please cite:\n",
      "\n",
      "LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package for phylogenetic tree input and output with richly annotated and associated data. Molecular Biology and Evolution 2020, 37(2):599-603. doi: 10.1093/molbev/msz240\n",
      "\n",
      "\n",
      "ggtree v2.3.7  For help: https://yulab-smu.github.io/treedata-book/\n",
      "\n",
      "If you use ggtree in published research, please cite the most appropriate paper(s):\n",
      "\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu. Using ggtree to visualize data on tree-like structures. Current Protocols in Bioinformatics, 2020, 69:e96. doi:10.1002/cpbi.96\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution 2018, 35(12):3041-3043. doi:10.1093/molbev/msy194\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36. doi:10.1111/2041-210X.12628\n",
      "\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ───────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.2     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.0.4     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ──────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m masks \u001b[34mggtree\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "\n",
      "Attaching package: ‘ggstance’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:ggplot2’:\n",
      "\n",
      "    geom_errorbarh, GeomErrorbarh\n",
      "\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggtree’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggtree’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘XVector’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:treeio’:\n",
      "\n",
      "    mask\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘DT’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:shiny’:\n",
      "\n",
      "    dataTableOutput, renderDataTable\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(treeio)\n",
    "library(ggtree)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(ggstance)\n",
    "library(Biostrings)\n",
    "library(ggmsa)\n",
    "library(seqmagick)\n",
    "library(XVector)\n",
    "library(gtable)\n",
    "library(grid)\n",
    "library(shiny)\n",
    "library(DT)\n",
    "library(cowplot)\n",
    "library(ggplotify)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Likely, the simplest shiny app ever"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Listening on http://127.0.0.1:7254\n",
      "\n"
     ]
    }
   ],
   "source": [
    "server <- function(input, output, session) { } #the server\n",
    "\n",
    "ui <- basicPage(\"This is a real Shiny app and jupyter is the best thing ever ;D\") # the user interface\n",
    "\n",
    "shinyApp(ui = ui, server = server) # this launches your app"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ok so let's build a web app which displays some trees\n",
    "## overview: 4 parts - preprocessing, ui, server, app calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## define my two classes - defining the A2TEA HYPOTHESES object ;D\n",
    "# class for the expanded_OG - containing all different types of data we have on it\n",
    "setClass(\"expanded_OG\", slots=list(genes=\"spec_tbl_df\", \n",
    "                                  fasta_files=\"list\", \n",
    "                                  msa=\"AAMultipleAlignment\", \n",
    "                                  tree=\"phylo\"))\n",
    "\n",
    "# class for the hypotheses\n",
    "setClass(\"hypothesis\", slots=list(description=\"character\", \n",
    "                                  number=\"character\",\n",
    "                                  expanded_in =\"character\", \n",
    "                                  compared_to=\"character\", \n",
    "                                  expanded_OGs=\"list\",\n",
    "                                  species_tree=\"phylo\"))\n",
    "\n",
    "\n",
    "#load the A2TEA input\n",
    "#load(\"/home/tyll/Desktop/PhD/A2TEA/shiny/example_trees/A2TEA_finished.RData\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ui"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define UI for tree app ----\n",
    "\n",
    "#hypothesesList <- as.list(hypotheses$hypothesis)\n",
    "#names(hypothesesList) <- hypotheses$name\n",
    "#hypothesesList\n",
    "\n",
    "\n",
    "# Allow specific errors to be displayed on screen, instead of displaying a generic error\n",
    "options(shiny.sanitize.errors = FALSE)\n",
    "\n",
    "ui <- pageWithSidebar(\n",
    "  \n",
    "  # App title ----\n",
    "  headerPanel(\"Tree building\"),\n",
    "  \n",
    "  # Sidebar panel for inputs ----\n",
    "  sidebarPanel(\n",
    "  \n",
    "      fileInput(inputId = \"A2TEA\",\n",
    "                label = \"Upload A2TEA.RData file:\",\n",
    "                multiple = FALSE,\n",
    "                accept = \".RData\"),\n",
    "      \n",
    "        # Input: alternative way which leads to problems...\n",
    "#      selectInput(\"select\",\n",
    "#                    label = h3(\"Select hypothesis:\"),\n",
    "#                    \"\"), \n",
    "\n",
    "      uiOutput('select_Hypothesis'),\n",
    "      uiOutput('select_HOG')\n",
    "      \n",
    "  ),\n",
    "  \n",
    "  # Main panel for displaying outputs ----\n",
    "  mainPanel(\n",
    "      h3(textOutput(\"text\")),\n",
    "      plotOutput(\"speciesTree\"),\n",
    "      plotOutput(\"VennUpSet_plot\"),\n",
    "      DT::dataTableOutput(\"HOG_DE_DT\"),\n",
    "      DT::dataTableOutput(\"HOG_level\"),\n",
    "      plotOutput(\"recFacetPlot\"),\n",
    "  )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#increasing maximum file size for upload!\n",
    "#https://stackoverflow.com/questions/18037737/how-to-change-maximum-upload-size-exceeded-restriction-in-shiny-and-save-user\n",
    "options(shiny.maxRequestSize=50*1024^2)\n",
    "\n",
    "myDataFrame <- data.frame(names=c(\"\"))\n",
    "#rm(myDataFrame)\n",
    "\n",
    "server <- function(input, output, session) {\n",
    "    \n",
    "\n",
    "#read-in hypotheses.tsv reactively\n",
    "    tsv <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t)\n",
    "    d <- get(\"hypotheses\", envir=t)\n",
    "     }\n",
    "     d\n",
    "   })  \n",
    "    \n",
    "#    observe({\n",
    "#       updateSelectInput(session, \"select\",\n",
    "#                         label = \"Select Hypothesis:\",\n",
    "#                         choices = tsv()$name,\n",
    "#                         selected = tsv()$name[1])\n",
    "# })\n",
    "    \n",
    "    \n",
    "#    output$hypothesesList <- renderText({\n",
    "#        tsv()$name\n",
    "#    })\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "#    tsv2 <- reactive({\n",
    "#       vars <- all.vars(parse(text = names(HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs)))\n",
    "#       vars <- as.list(vars)\n",
    "#       return(vars)\n",
    "    \n",
    "    \n",
    "    output$select_Hypothesis = renderUI({\n",
    "      selectInput('select_Hypothesis2', 'Select Hypothesis:', tsv()$name)\n",
    "    })\n",
    "\n",
    "# also read-in the Ortho Venn/UpSet plots \n",
    "    Ortho_intersect_plots <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t5 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t5)\n",
    "    d <- get(\"Ortho_intersect_plots\", envir = t5)\n",
    "     }\n",
    "     d\n",
    "   }) \n",
    "\n",
    "    #plot the Ortho Venn/UpSet plots\n",
    "    output$VennUpSet_plot <- renderPlot({\n",
    "        Ortho_intersect_plots()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]\n",
    "    })\n",
    "\n",
    "    \n",
    "# also read-in the DE genes table \n",
    "    HOG_DE <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t2 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t2)\n",
    "    d <- get(\"HOG_DE.a2tea\", envir = t2)\n",
    "     }\n",
    "     d\n",
    "   }) \n",
    "\n",
    "\n",
    "# also read-in custom A2TEA object\n",
    "    HYPOTHESES <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t3 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t3)\n",
    "    d <- get(\"HYPOTHESES.a2tea\", envir = t3)\n",
    "     }\n",
    "     d\n",
    "   })  \n",
    "\n",
    "# buld species tree\n",
    "    #plot the Ortho Venn/UpSet plots\n",
    "    output$speciesTree <- renderPlot({\n",
    "       st <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@species_tree\n",
    "       ggtree(st, layout = 'rectangular', branch.length='yes', size=1.5) +\n",
    "           geom_tiplab(align=F, offset = 0.1, alpha=1)\n",
    "            \n",
    "    })\n",
    "    \n",
    "    \n",
    "# also read-in hypothesis specific HOG level table - HOG_level_list\n",
    "    HOG_level <- reactive({\n",
    "     inFile <- input$A2TEA\n",
    "     if (is.null(inFile)) {\n",
    "         d <- myDataFrame\n",
    "    } else {\n",
    "    t4 = new.env()\n",
    "    load(input$A2TEA$datapath, envir = t4)\n",
    "    d <- get(\"HOG_level_list\", envir = t4)\n",
    "     }\n",
    "     d\n",
    "   })  \n",
    "    \n",
    "    \n",
    "# normal table chokes on large size of df\n",
    "# therefore usage of DT::renderDataTable\n",
    "    output$HOG_DE_DT <- DT::renderDataTable({\n",
    "        HOG_DE()\n",
    "  })\n",
    "    \n",
    "    \n",
    "#example access to hypothesis number - useful for access \n",
    "    hypothesisNUM <- reactive({\n",
    "#       HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@number\n",
    "        HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@number\n",
    "    })\n",
    "   \n",
    "     output$text <- renderText({\n",
    "        hypothesisNUM()\n",
    "    })    \n",
    "        \n",
    "        \n",
    "\n",
    "        \n",
    "# I can make use of the hypothesis number previously created\n",
    "    output$HOG_level <- DT::renderDataTable({\n",
    "#        HOG_level()[[tsv()$hypothesis[tsv()$name==input$select]]]\n",
    "                HOG_level()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]\n",
    "  })\n",
    "             \n",
    "     \n",
    "        \n",
    "\n",
    "\n",
    "# observe choice of HOG; e.g. for building trees\n",
    "# first create reactive list of all exp. HOGs \n",
    "# due to my custom object this needs the following trickery: \n",
    "# all.vars(parse(text = names(....\n",
    "hypothesisExpOGs <- reactive({\n",
    "#       vars <- all.vars(parse(text = names(HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs)))\n",
    "       vars <- all.vars(parse(text = names(HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@expanded_OGs)))    \n",
    "       vars <- as.list(vars)\n",
    "       return(vars)    \n",
    "    })\n",
    "    \n",
    "# actually render the UI element - a dropdown list of the expanded HOGs of the current hypothesis\n",
    "    output$select_HOG = renderUI({\n",
    "      selectInput('select_HOG2', 'Select HOG', hypothesisExpOGs())\n",
    "    })\n",
    "    \n",
    "# creating the rectangular A2TEA triple facet plot\n",
    "\n",
    "# make renderUI element reactive\n",
    "# access to element based on name in first element selectInput('')\n",
    "    choice <- reactive({input$select_HOG2})\n",
    "    \n",
    "output$recFacetPlot <- renderPlot({\n",
    "\n",
    "    #### building the rectangular facet plot - HOG tree, logFC & MSA\n",
    "                     \n",
    "#    tree <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs[[choice()]]@tree\n",
    "    tree <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@expanded_OGs[[choice()]]@tree\n",
    "    info <- HOG_DE()\n",
    "\n",
    "\n",
    "# extracted from colourblind - removed black... how to deal with NAs..\n",
    "# also a problem if more than given colours are supposed to be displayed error....\n",
    "# could perform check of number of HOGs and switch between an optimized colour scale with predefined colours\n",
    "# to an open-ended/automatic colour scheme if set number is exceeded by choice of HOG\n",
    "    cols <- c(\"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\", \"#D55E00\", \"#CC79A7\", \"lightblue\", \"darkred\",\n",
    "              \"lightblue\", \"darkred\", \"pink\", \"cyan\", \"grey\", \"violet\", \"magenta\", \"gold\", \n",
    "              \"darkgreen\", \"darkblue\")\n",
    "\n",
    "\n",
    "# rlang package has the function is_empty()\n",
    "# with it we can test for \"character(0)\"\n",
    "    el <- list()\n",
    "\n",
    "    for (i in 1:length(tree$tip.label)) {\n",
    "        if (is_empty(info$HOG[info$gene == tree$tip.label[i]]) == FALSE) {\n",
    "           first <- info$HOG[info$gene == tree$tip.label[i]]\n",
    "           el <- c(el, list(filler = tree$tip.label[i]))\n",
    "           names(el)[i] <- first\n",
    "        }\n",
    "        else {\n",
    "            el <- c(el, list(singleton = tree$tip.label[i]))      \n",
    "        }\n",
    "    }\n",
    "\n",
    "# reduce the list to unique tags\n",
    "    el_split <- sapply(unique(names(el)), function(x) unname(unlist(el[names(el)==x])), simplify=FALSE)\n",
    "\n",
    "    p_OTU <- ggtree(tree,\n",
    "                    layout = 'rectangular',\n",
    "                    branch.length='none',\n",
    "                    size=1.5)\n",
    "\n",
    "    p <- groupOTU(p_OTU, el_split, 'HOG') + \n",
    "                       aes(color=HOG) +\n",
    "                       geom_tippoint(aes(color=HOG)) +\n",
    "                       scale_color_manual(values=cols,\n",
    "                                          na.translate=TRUE,\n",
    "                                          na.value = \"black\") +\n",
    "                       geom_tiplab(aes(color=HOG), \n",
    "                                   align=F, \n",
    "                                   offset = 0.3,\n",
    "                                   alpha=1)\n",
    "                                   \n",
    "\n",
    "# this is such a convenient way to do this ;D + we also drop the columns we don't need\n",
    "    d <- filter(info, gene %in% tree$tip.label) %>% \n",
    "             select(-c(species, baseMean, HOG, lfcSE, stat, pvalue, padj))\n",
    "\n",
    "    lines=data.frame(y = c(-2,-1,1,2), .panel='log2FC')\n",
    "\n",
    "                   \n",
    "# xlim set x axis limits for only Tree panel\n",
    "# xlim tree calculated as number of tips tips?\n",
    "    p2 <- facet_plot(p + xlim_tree(16), \n",
    "                     panel = 'log2FC', \n",
    "                     data = d, \n",
    "                     geom = geom_barh, \n",
    "                     mapping = aes(x = log2FoldChange),\n",
    "                     stat='identity',\n",
    "                 #inherit.aes = TRUE,\n",
    "                     color='firebrick') + \n",
    "                       theme_tree2(plot.margin=margin(0, 0, 0, 0)) + \n",
    "                       geom_vline(data=lines, \n",
    "                                  aes(xintercept=y), \n",
    "                                  linetype = \"dashed\")\n",
    "                   \n",
    "                   \n",
    "# get the msa and assign it to x (for now...)\n",
    "# set msa range of amino acids to be displayed - make this dynamcic!\n",
    "#    x <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select]]]@expanded_OGs[[choice()]]@msa\n",
    "    x <- HYPOTHESES()[[tsv()$hypothesis[tsv()$name==input$select_Hypothesis2]]]@expanded_OGs[[choice()]]@msa\n",
    "    data <- tidy_msa(x, start = 1, end = 598)\n",
    "\n",
    "\n",
    "# use font to display amino acid code, eg \"helvetical\"\n",
    "p3 <- p2 + geom_facet(geom = geom_msa, \n",
    "               data = data,  \n",
    "               panel = 'msa',\n",
    "               font = NULL, \n",
    "               color = \"Clustal\")\n",
    "                   \n",
    "\n",
    "    gt = ggplot_gtable(ggplot_build(p3))\n",
    "#gt$layout$l[grep('background', gt$layout$name)]\n",
    "#gt$widths[1] = 20*gt$widths[1] # in this case it was colmun 5 - double the width\n",
    "    gt$widths[5] = 2*gt$widths[5] # in this case it was colmun 5 - double the width\n",
    "    gt$widths[7] = 2*gt$widths[7] # in this case it was colmun 7 - double the width\n",
    "    gt$widths[9] = 5*gt$widths[9] # in this case it was colmun 9 - double the width\n",
    "\n",
    "# this option is nice for jupyter but I have to see what will be possible in shiny                   \n",
    "    options(repr.plot.width=15, repr.plot.height=8)\n",
    "    grid.draw(gt)\n",
    "            \n",
    "            \n",
    "############ ending facet plot code ############\n",
    "            \n",
    "  })\n",
    "        \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### app calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Listening on http://127.0.0.1:3107\n",
      "\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in .subset2: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“Error in [[: attempt to select less than one element in get1index”\n",
      "Warning message:\n",
      "“`filter_()` is deprecated as of dplyr 0.7.0.\n",
      "Please use `filter()` instead.\n",
      "See vignette('programming') for more help\n",
      "\u001b[90mThis warning is displayed once every 8 hours.\u001b[39m\n",
      "\u001b[90mCall `lifecycle::last_warnings()` to see where this warning was generated.\u001b[39m”\n",
      "Warning message:\n",
      "“Removed 3 rows containing missing values (position_stack).”\n"
     ]
    }
   ],
   "source": [
    "shinyApp(ui, server)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### playground & working examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2"
      ],
      "text/latex": [
       "2"
      ],
      "text/markdown": [
       "2"
      ],
      "text/plain": [
       "[1] 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hypotheses$hypothesis[hypotheses$name==\"Expanded in barley compared to maize\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS( data.frame(names=c(\"Jill\",\"Jane\",\"Megan\")),\"example_trees/myDataFrame.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### completely working mini application showcasing runtime loading of data and reactive dropdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Listening on http://127.0.0.1:5402\n",
      "\n"
     ]
    }
   ],
   "source": [
    "myDataFrame <- data.frame(names=c(\"Tom\",\"Dick\",\"Harry\"))\n",
    "\n",
    "ui <- shinyUI(\n",
    "    fluidPage(\n",
    "        fileInput(\"file1\", \"Choose file to upload\", accept = \".rds\"),\n",
    "        selectInput(\"myNames\",\"Names\", \"\"),\n",
    "        tableOutput(\"contents\")\n",
    "    )\n",
    ")\n",
    "\n",
    "server <- function(input, output, session) {\n",
    "\n",
    "    myData <- reactive({\n",
    "        inFile <- input$file1\n",
    "        if (is.null(inFile)) {\n",
    "            d <- myDataFrame\n",
    "        } else {\n",
    "            d <- readRDS(inFile$datapath)\n",
    "        }\n",
    "        d\n",
    "    })\n",
    "\n",
    "    output$contents <- renderTable({\n",
    "        myData()\n",
    "    })\n",
    "\n",
    "    observe({\n",
    "         updateSelectInput(session, \"myNames\",\n",
    "                           label = \"myNames\",\n",
    "                           choices = myData()$names,\n",
    "                           selected = myData()$names[1])\n",
    "    })\n",
    "\n",
    "}\n",
    "\n",
    "shinyApp(ui, server)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### shiny example of using RData input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Listening on http://127.0.0.1:4827\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#The example allows you to upload .RData files.\n",
    "#The approach with load and get allows you to assign the loaded data to a variable name of your choice.\n",
    "#For the matter of the example being \"standalone\" I inserted the top section that stores \n",
    "#two vectors to your disk in order to load and plot them later.\n",
    "\n",
    "\n",
    "# Define two datasets and store them to disk\n",
    "x <- rnorm(100)\n",
    "save(x, file = \"x.RData\")\n",
    "#rm(x)\n",
    "y <- rnorm(100, mean = 2)\n",
    "save(y, file = \"y.RData\")\n",
    "#rm(y)\n",
    "# both\n",
    "save(x, y, file = \"z.RData\")\n",
    "\n",
    "# This function, borrowed from http://www.r-bloggers.com/safe-loading-of-rdata-files/, load the Rdata into a new environment to avoid side effects\n",
    "LoadToEnvironment <- function(RData, env=new.env()) {\n",
    "  load(RData, env)\n",
    "  return(env)\n",
    "}\n",
    "\n",
    "\n",
    "# Define UI\n",
    "ui <- shinyUI(fluidPage(\n",
    "  titlePanel(\".RData File Upload Test\"),\n",
    "  mainPanel(\n",
    "    fileInput(\"file\", label = \"\"),\n",
    "    actionButton(inputId=\"plot\",\"Plot\"),\n",
    "    plotOutput(\"hist\"))\n",
    "  )\n",
    ")\n",
    "\n",
    "# Define server logic\n",
    "server <- shinyServer(function(input, output) {\n",
    "\n",
    "  observeEvent(input$plot,{\n",
    "    if ( is.null(input$file)) return(NULL)\n",
    "    inFile <- input$file\n",
    "    file <- inFile$datapath\n",
    "    # load the file into new environment and get it from there\n",
    "#    name <- load(file, envir = e)\n",
    "#    data <- e[[name]]\n",
    "    # this works but not really what I need\n",
    "#    data <- e[[name[1]]]\n",
    "    t = new.env()\n",
    "    load(file, envir = t)\n",
    "    data <- get(\"y\", envir=t)\n",
    "\n",
    "      \n",
    "      \n",
    "#    observeEvent(input$plot,{\n",
    "#        if ( is.null(input$file)) return(NULL)\n",
    "#        inFile <- input$file\n",
    "#        file <- inFile$datapath\n",
    "#    # Use a reactiveFileReader to read the file on change, and load the content into a new environment\n",
    "#    env <- reactiveFileReader(1000, session, \"z.RData\", LoadToEnvironment)\n",
    "#    # Access the first item in the new environment, assuming that the Rdata contains only 2 items which is a data frame\n",
    "#    env()[[names(env())[2]]]\n",
    "    \n",
    "    \n",
    "    # Plot the data\n",
    "    output$hist <- renderPlot({\n",
    "      hist(data)\n",
    "    })\n",
    "  })\n",
    "})\n",
    "\n",
    "# Run the application \n",
    "shinyApp(ui = ui, server = server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " num [1:100] 3.011 0.694 1.618 1.81 2.27 ...\n",
      " num [1:100] 3.011 0.694 1.618 1.81 2.27 ...\n"
     ]
    }
   ],
   "source": [
    "e = new.env()\n",
    "name <- load(\"z.RData\", envir = e)\n",
    "#mget(c(\"y\",\"x\"), envir=e)\n",
    "#get returns exactly the object\n",
    "test <- get(\"y\", envir=e)\n",
    "str(test)\n",
    "str(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             Length Class      Mode\n",
       "hypothesis_1 1      hypothesis S4  \n",
       "hypothesis_2 1      hypothesis S4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A spec_tbl_df: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>hypothesis</th><th scope=col>name</th><th scope=col>expanded_in</th><th scope=col>compared_to</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>Expanded in Arabidopsis compared to Monocots</td><td>Arabidopsis_thaliana</td><td>Zea_mays;Hordeum_vulgare</td></tr>\n",
       "\t<tr><td>2</td><td>Expanded in barley compared to maize        </td><td>Hordeum_vulgare     </td><td>Zea_mays                </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A spec\\_tbl\\_df: 2 × 4\n",
       "\\begin{tabular}{llll}\n",
       " hypothesis & name & expanded\\_in & compared\\_to\\\\\n",
       " <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & Expanded in Arabidopsis compared to Monocots & Arabidopsis\\_thaliana & Zea\\_mays;Hordeum\\_vulgare\\\\\n",
       "\t 2 & Expanded in barley compared to maize         & Hordeum\\_vulgare      & Zea\\_mays                \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A spec_tbl_df: 2 × 4\n",
       "\n",
       "| hypothesis &lt;dbl&gt; | name &lt;chr&gt; | expanded_in &lt;chr&gt; | compared_to &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | Expanded in Arabidopsis compared to Monocots | Arabidopsis_thaliana | Zea_mays;Hordeum_vulgare |\n",
       "| 2 | Expanded in barley compared to maize         | Hordeum_vulgare      | Zea_mays                 |\n",
       "\n"
      ],
      "text/plain": [
       "  hypothesis name                                         expanded_in         \n",
       "1 1          Expanded in Arabidopsis compared to Monocots Arabidopsis_thaliana\n",
       "2 2          Expanded in barley compared to maize         Hordeum_vulgare     \n",
       "  compared_to             \n",
       "1 Zea_mays;Hordeum_vulgare\n",
       "2 Zea_mays                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "             Length Class      Mode\n",
       "hypothesis_1 1      hypothesis S4  \n",
       "hypothesis_2 1      hypothesis S4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "load(\"example_trees/backup_old_RData//A2TEA_finished.RData\")\n",
    "#str(hypotheses$name)\n",
    "#HOG_DE.a2tea\n",
    "#HYPOTHESES.a2tea\n",
    "#HOG_level_list[[1]][HOG_level_list[[1]]$expansion==\"yes\",]\n",
    "#tt <- as.data.frame(names(HYPOTHESES.a2tea[[1]]@expanded_OGs))\n",
    "#names(tt)\n",
    "#names(tt)[names(tt) == \"names(HYPOTHESES.a2tea[[1]]@expanded_OGs)\"] <- \"HOG\"\n",
    "#names(tt)\n",
    "#tt$HOG\n",
    "#tt[[1]]\n",
    "summary(HYPOTHESES.a2tea)\n",
    "#?base::as.data.frame\n",
    "hypotheses\n",
    "\n",
    "\n",
    "load(\"example_trees/A2TEA_finished.RData\")\n",
    "#str(HOG_level_list)\n",
    "#hypotheses\n",
    "summary(HYPOTHESES.a2tea)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
